routers/private/hook_pre_receive.go

```diff
@@ -439,6 +442,51 @@ func preReceivePullRequest(ctx *preReceiveContext, oldCommitID, newCommitID, ref
 	}
 }
 
+func preReceiveDeletePRHeadRef(ctx *preReceiveContext, refFullName string) {
+	if !ctx.opts.GitPushOptions.Bool("delete-pull-head-confirm", false) {
+		ctx.JSON(http.StatusForbidden, private.Response{
+			Err: "Warning: you are attempting to delete the head reference of a pull request: " + refFullName + ".\n" +
+				"PR head references can only be deleted for closed PRs and only by their author or the repository admin.\n" +
+				"If you are sure what you want to do this, push again with the push option `git push ... -o delete-pull-head-confirm=true` ",
+		})
+		return
+	}
+
+	pullIndexStr := strings.TrimPrefix(refFullName, git.PullPrefix)
+	pullIndexStr = strings.Split(pullIndexStr, "/")[0]
+	pullIndex, _ := strconv.ParseInt(pullIndexStr, 10, 64)
+	if pullIndex <= 0 {
+		return
+	}
+
+	pr, err := models.GetPullRequestByIndex(ctx.Repo.Repository.ID, pullIndex)
+	if err != nil {
+		log.Error("Unable to get PR #%d in %s Error: %v", pullIndex, ctx.Repo.Repository.FullName(), err)
+		ctx.JSON(http.StatusInternalServerError, private.Response{
+			Err: fmt.Sprintf("Unable to get PR #%d Error: %v", pullIndex, err),
+		})
+		return
+	}
+
+	if !pr.Issue.IsClosed {
+		ctx.JSON(http.StatusForbidden, private.Response{
+			Err: fmt.Sprintf("Error: Head references of open pull request: %s cannot be deleted.", refFullName),
+		})
+		return
+	}
+
+	if pr.Issue.PosterID == ctx.opts.UserID {
```

Comment:
I still don't like that PR creator would be able to do this as in some cases this could go against some policy and there is no way to prevent this.

Reply:
Reason? I think it's fine, I can not think about a case the it would destroy the target repository or go against some policy.

Do you have examples and cases?

Reply:
Like I'm as the repository owner would not be able to see PR's changes anymore ;)

Reply:
But if the author deletes the PR, they must be very unwilling to let you see. Isn't it correct?

Reply:
Gitea does not allow deleting PR's 😉 It still should be up to the owner to decide what to delete and what not as this is his repository

Reply:
Although the PR belongs to the repo, the owner of the PR is the author (the real owner). And I can imagine a case: if the PR author pushed some sensitive data (password), without deleting the branch, does they have other choices? If yes, I am fine to either (allow deleting or not). If no, I think the author should be able to delete the branch.

Not sure whether my understanding is right, correct me if I am wrong.

Reply:
Yes, owner can delete his branch but I would like to prevent some malicious user deleting for example all PRs heads for already merged PRs

Reply:
> Yes, owner can delete his branch b

But what if the owner doesn't want to help? For example, the owner is offline for a long time.

>  prevent some malicious user deleting for example all his PRs heads for already merged PRs

It's another check if I understand correctly.

Reply:
I meant to maliciously delete all his own PRs heads that are already merged

Reply:
It's beyond my comprehension. Maybe it's worth for a poll

Reply:
> I meant to maliciously delete all his own PRs heads that are already merged

done. https://github.com/go-gitea/gitea/pull/18208/commits/7912a2bfb7e785e100f56acfab1be8debaa096e1